File: library/src/main/java/cn/zfs/widget/TouchControlableViewPager.java
package cn.zfs.widget;

import android.content.Context;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.animation.Interpolator;
import android.view.animation.OvershootInterpolator;
import android.widget.Scroller;

import java.lang.reflect.Field;

/**
 * Created by zengfs on 2016/2/2.
 * 可控制是否能滑动
 */
public class TouchControlableViewPager extends ViewPager {
	
	private boolean isTouchEnabled = true;
	
	public TouchControlableViewPager(Context context) {
		super(context);
	}

	public TouchControlableViewPager(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	/**
	 * 设置ViewPager是否可滑动
	 */
	public void setTouchEnabled(boolean enable) {
		isTouchEnabled = enable;
	}
	
	/**
	 * 是否禁用其事件，实现不能滑动
	 */
	@Override
	public boolean onTouchEvent(MotionEvent ev) {
		return isTouchEnabled && super.onTouchEvent(ev);
	}

	/**
	 * 是否截触摸事件，让事件可以向内层传递
	 */
	@Override
	public boolean onInterceptTouchEvent(MotionEvent ev) {
		return isTouchEnabled && super.onInterceptTouchEvent(ev);
	}

    /**设置切换动画时长*/
    public void setScrollDuration(int duration) {
        try {
            Field field = ViewPager.class.getDeclaredField("mScroller");
            field.setAccessible(true);
            ViewPagerScroller viewPagerScroller = new ViewPagerScroller(getContext(), new OvershootInterpolator(0.6F));
            field.set(this, viewPagerScroller);
            viewPagerScroller.setDuration(duration);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 控制ViewPager动画切换时长
     */
    public class ViewPagerScroller extends Scroller {
        private int duration;

        public ViewPagerScroller(Context context) {
            super(context);
        }

        public ViewPagerScroller(Context context, Interpolator interpolator) {
            super(context, interpolator);
        }

        public void setDuration(int duration) {
            this.duration = duration;
        }

        @Override
        public void startScroll(int startX, int startY, int dx, int dy) {
            super.startScroll(startX, startY, dx, dy, duration);
        }

        @Override
        public void startScroll(int startX, int startY, int dx, int dy, int duration) {
            super.startScroll(startX, startY, dx, dy, this.duration);
        }
    }
}

File: library/src/main/java/com/snail/widget/TouchControlableViewPager.kt
package com.snail.widget

import android.content.Context
import android.support.v4.view.ViewPager
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.animation.Interpolator
import android.view.animation.OvershootInterpolator
import android.widget.Scroller

/**
 * Created by zengfs on 2016/2/2.
 * 可控制是否能滑动
 */
class TouchControlableViewPager @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : ViewPager(context, attrs) {

    private var isTouchEnabled = true

    /**
     * 设置ViewPager是否可滑动
     */
    fun setTouchEnabled(enable: Boolean) {
        isTouchEnabled = enable
    }

    /**
     * 是否禁用其事件，实现不能滑动
     */
    override fun onTouchEvent(ev: MotionEvent): Boolean {
        return isTouchEnabled && super.onTouchEvent(ev)
    }

    /**
     * 是否截触摸事件，让事件可以向内层传递
     */
    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        return isTouchEnabled && super.onInterceptTouchEvent(ev)
    }

    /**设置切换动画时长 */
    fun setScrollDuration(duration: Int) {
        try {
            val field = ViewPager::class.java.getDeclaredField("mScroller")
            field.isAccessible = true
            val viewPagerScroller = ViewPagerScroller(context, OvershootInterpolator(0.6f))
            field.set(this, viewPagerScroller)
            viewPagerScroller.duration = duration
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 控制ViewPager动画切换时长
     */
    inner class ViewPagerScroller @JvmOverloads constructor(context: Context, interpolator: Interpolator? = null) : Scroller(context, interpolator) {
        internal var duration: Int = 0

        fun setDuration(duration: Int) {
            this.duration = duration
        }

        override fun startScroll(startX: Int, startY: Int, dx: Int, dy: Int) {
            super.startScroll(startX, startY, dx, dy, duration)
        }

        override fun startScroll(startX: Int, startY: Int, dx: Int, dy: Int, duration: Int) {
            super.startScroll(startX, startY, dx, dy, this.duration)
        }
    }
}
